### PR TITLE
Replace Ref by AtomicReference

### PR DESCRIPTION
**Pull request by rtar**

### CODE REVIEW

File: actor/src/main/scala/com/evolutiongaming/akkaeffect/util/Serially.scala [hunk 1]
```diff
@@ -44,27 +57,33 @@ private[akkaeffect] object Serially {
...
-                for {
-                  a <- s.task(a)
-                  a <- t(a)
-                } yield a
-              }
-              (S.Active(task), Concurrent[F].unit)
-            case S.Active    =>
-              (S.Active(t), Concurrent[F].unit)
+          a <- Sync[F].delay {
+            val s = ref.getAndUpdate {
```
Comment (t3hnar) [inline actor/src/main/scala/com/evolutiongaming/akkaeffect/util/Serially.scala:69]: `s` is a previous state here, how is it supposed to work?

File: actor/src/main/scala/com/evolutiongaming/akkaeffect/util/Serially.scala [hunk 2]
```diff
@@ -44,27 +57,33 @@ private[akkaeffect] object Serially {
...
-              val task = (a: A) => Async[F].defer {
-                for {
-                  a <- s.task(a)
-                  a <- t(a)
-                } yield a
-              }
-              (S.Active(task), Concurrent[F].unit)
-            case S.Active    =>
-              (S.Active(t), Concurrent[F].unit)
+          a <- Sync[F].delay {
```
Comment (t3hnar) [inline actor/src/main/scala/com/evolutiongaming/akkaeffect/util/Serially.scala:68]: no need to write everything in `Sync[F].delay {`

**Review by t3hnar — CHANGES_REQUESTED** [review]

File: actor/src/main/scala/com/evolutiongaming/akkaeffect/util/Serially.scala (hunk 1 above)
Comment (rtar) [reply actor/src/main/scala/com/evolutiongaming/akkaeffect/util/Serially.scala:69]: It is only used to select a return value, the same as in original version.

`AtomicReference` does not have `Ref.modify`, so one either have to write spin loop himself, or write "match...case" twice together with `getOrUpdate`.

Comment (t3hnar) [reply actor/src/main/scala/com/evolutiongaming/akkaeffect/util/Serially.scala:69]: unfortunately one should write spin loop himself…

Comment (rtar) [reply actor/src/main/scala/com/evolutiongaming/akkaeffect/util/Serially.scala:69]: Why the current version is broken? I do not see the bug somehow...

It updates a state to a new one and then uses the old one to select the second part of the tuple. It looks like a full equivalent of the custom spin loop. 🤔

Comment (t3hnar) [reply actor/src/main/scala/com/evolutiongaming/akkaeffect/util/Serially.scala:69]: I checked once more - I was wrong, it will work!

**Review by t3hnar — CHANGES_REQUESTED** [review]

File: actor/src/main/scala/com/evolutiongaming/akkaeffect/util/Serially.scala [hunk 3]
```diff
@@ -22,18 +28,25 @@ private[akkaeffect] object Serially {
...
           a <- task(a)
-          a <- ref.modify {
-            case s: S.Active => (S.Active, (a, s.task).asLeft[Unit])
-            case S.Active    => (S.Idle(a), unit)
-            case _: S.Idle   => (S.Idle(a), unit)
+          a <- Sync[F].delay {
+            val s = ref.getAndUpdate {
+              case _: S.Active => S.Active
+              case S.Active    => S.Idle(a)
+              case _: S.Idle   => S.Idle(a)
```
[the remark this answers is not in the document]
Comment (rtar) [reply actor/src/main/scala/com/evolutiongaming/akkaeffect/util/Serially.scala:43]: Should we really throw out the new value here? :thinking:

File: actor/src/main/scala/com/evolutiongaming/akkaeffect/util/Serially.scala (hunk 2 above)
Comment (rtar) [reply actor/src/main/scala/com/evolutiongaming/akkaeffect/util/Serially.scala]: Fixed

**Review by t3hnar — APPROVED** [review]

File: actor/src/main/scala/com/evolutiongaming/akkaeffect/util/Serially.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (t3hnar) [reply actor/src/main/scala/com/evolutiongaming/akkaeffect/util/Serially.scala:43]: you can throw out exception here :)

**Pull request merged**